go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -212,18 +351,81 @@ func (npw *nodePortWatcher) SyncServices(services []interface{}) {
+		npw.updateServiceFlowCache(service, true, false)
 	}
 
 	npw.ofm.requestFlowSync()
-	syncSharedGatewayIptRules(services)
+	syncSharedGatewayIptRules(services, false, npw.nodeIPManager.addresses.List())
+}
+
+// For endpoint Event functions and exernalTrafficPolicy:local
+//     An action only ever occurs (i.e epHostLocal=True) if
+//			1. The Service Exists &&
+// 			2. Svc externalTrafficpolicy:local &&
+//			3. Svc Has host networked Backends
+//      Otherwise the endpoint event is ignored
+
+// Add new Local Traffic flows if endpoints are local
+func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
+	var svc interface{}
+	var err error
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+
+	// Endpoint add Event can arrive before service add event, get and create it if it doesn't exist
+	svc, exists := npw.services.Load(name)
+	if !exists {
+		klog.V(5).Infof("Endpoint: %s event in namespace: %s arrived before service creation, create service: %s",
+			ep.Name, ep.Namespace, ep.Name)
+		svc, err = npw.watchFactory.GetService(ep.Namespace, ep.Name)
+		if err != nil {
+			// This is not necessarily an error. For e.g when there are endpoints
+			// without a corresponding service.
+			klog.V(5).Infof("No service found for endpoint %s in namespace %s", ep.Name, ep.Namespace)
+			return
+		}
+		npw.AddService(svc.(*kapi.Service))
+	}
+
+	if countHostNeworkEndpoints(ep, &npw.nodeIPManager.addresses) && util.ServiceExternalTrafficPolicyLocal(svc.(*kapi.Service)) {
+		klog.V(5).Infof("Service %s has ExternalTrafficPolicy=local and local Endpoints, Updating flow Cache", ep.Name)
+		// Update Service Cache with special flows
+		npw.updateServiceFlowCache(svc.(*kapi.Service), true, true)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(svc.(*kapi.Service), false, npw.nodeIPManager.addresses.List())
+		addSharedGatewayIptRules(svc.(*kapi.Service), true, npw.nodeIPManager.addresses.List())
+	}
+}
+
+func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+	if svc, exists := npw.services.Load(name); exists && util.ServiceExternalTrafficPolicyLocal(svc.(*kapi.Service)) {
+		// Always try and delete flows/iptables rules here regardless of having host networked endpoints
+		npw.updateServiceFlowCache(svc.(*kapi.Service), false, true)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(svc.(*kapi.Service), true, npw.nodeIPManager.addresses.List())
+
+	}
+	npw.services.Delete(name)
+}
+
+func (npw *nodePortWatcher) UpdateEndpoints(old *kapi.Endpoints, new *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: new.Namespace, Name: new.Name}
+	// Service should always exist here
+	if svc, exists := npw.services.Load(name); exists && util.ServiceExternalTrafficPolicyLocal(svc.(*kapi.Service)) {
+		// only operate here if the endpoint's addresses have changed
+		if !reflect.DeepEqual(old.Subsets, new.Subsets) {
```

Comment:
this isnt correct, you need to check if old.Subsets had host endpoints, and if new also has host endpoints (then you dont need to do anything), otherwise you need to do some different logic. This logic makes delete and add happen for no reason

Reply:
Done, Now I only delete and re-add if 
- old and new are not equal AND
- ETP:local 

Then the Logic changes based on changes to the host networked endpoint state